tools/thanosconvert/thanosconvert.go

```diff
@@ -0,0 +1,195 @@
+		}
+	}
+	return results, nil
+}
+
+func (c ThanosBlockConverter) convertUser(ctx context.Context, user string) (PerUserResults, error) {
+	results := PerUserResults{}
+
+	err := c.bkt.Iter(ctx, user, func(o string) error {
+		// get block ID from path
+
+		parts := strings.Split(o, "/")
+		if len(parts) != 3 {
+			// this seems pretty bad, let's bail out
+			return errors.Errorf("couldn't parse bucket path %s as {user}/{block}/", o)
+		}
+		blockID := parts[1]
+		level.Debug(c.logger).Log("msg", "Processing block", "block", blockID)
+
+		// retrieve meta.json
+
+		metaKey := fmt.Sprintf("%s/%s/meta.json", user, blockID)
+		r, err := c.bkt.Get(ctx, metaKey)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "Get meta.json", "block", blockID, "meta_key", metaKey, "err", err.Error())
+			results.AddFailed(blockID)
+			return nil
+		}
+
+		meta := &metadata.Meta{}
+		decoder := json.NewDecoder(r)
+		err = decoder.Decode(meta)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "decode meta.json into metadata.Meta", "block", blockID, "meta_key", metaKey, "err", err.Error())
+			results.AddFailed(blockID)
+			return nil
+		}
+
+		// convert and upload if appropriate
+
+		newMeta, changesRequired := convertMetadata(*meta, user)
+
+		if len(changesRequired) > 0 {
+			if c.dryRun {
+				level.Debug(c.logger).Log("msg", "Block requires changes, not uploading due to dry run", "block", blockID, "changes_required", strings.Join(changesRequired, ","))
+			} else {
+				level.Debug(c.logger).Log("msg", "Block requires changes, uploading new meta.json", "block", blockID, "changes_required", strings.Join(changesRequired, ","))
+				if err := c.uploadNewMeta(ctx, user, blockID, newMeta); err != nil {
+					level.Error(c.logger).Log("msg", "Update meta.json", "block", blockID, "meta_key", metaKey, "err", err.Error())
+					results.AddFailed(blockID)
+					return nil
+				}
+			}
+			results.AddConverted(blockID)
+		} else {
+			results.AddUnchanged(blockID)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return results, err
+	}
+
+	return results, nil
+}
+
+func (c *ThanosBlockConverter) uploadNewMeta(ctx context.Context, user, blockID string, meta metadata.Meta) error {
+	key := fmt.Sprintf("%s/%s/meta.json", user, blockID)
+
+	var body bytes.Buffer
+	enc := json.NewEncoder(&body)
+	if err := enc.Encode(meta); err != nil {
+		return errors.Wrap(err, "encode json")
+	}
+
+	if err := c.bkt.Upload(ctx, key, &body); err != nil {
+		return errors.Wrap(err, "upload to bucket")
+	}
+	return nil
+}
+
+func convertMetadata(meta metadata.Meta, expectedUser string) (metadata.Meta, []string) {
+	var changesRequired []string
+
+	if meta.Thanos.Labels == nil {
+		// TODO: no "thanos" entry could mean this block is corrupt - should we bail out instead?
+		meta.Thanos.Labels = map[string]string{}
+		// don't need to add to changesRequired, since the code below will notice lack of org id
+	}
+
+	org, ok := meta.Thanos.Labels["__org_id__"]
```

Comment:
Instead of the hardcoded string `"__org_id__"`, please use `cortex_tsdb.TenantIDExternalLabel`.